src/managers/MilestoneEventManager.ts

```diff
@@ -0,0 +1,179 @@
+		this.checkMilestonesLazily();
+	}
+
+	private _onEventHandler(textDocChangeEvent: TextDocumentChangeEvent) {
+		// set a timer to check so we don't disrupt typing
+		this.checkMilestonesLazily();
+	}
+
+	private _windowStateChanged(winState: WindowState) {
+		this.checkMilestonesLazily();
+	}
+
+	private checkMilestonesLazily() {
+		if (this._checkMilestoneTimeout) {
+			// cancel the current one
+			clearTimeout(this._checkMilestoneTimeout);
+			this._checkMilestoneTimeout = null;
+		}
+		this._checkMilestoneTimeout = setTimeout(() => {
+			this.checkForMilestones();
+		}, 1000 * 30);
+	}
+
+	private async checkForMilestones() {
+		if (!isLoggedIn() || this._checkingForMilestones) {
+			return;
+		}
+		this._checkingForMilestones = true;
+		// updates logs with latest metrics and checks for milestones
+
+		// checks to see if there are any new achieved milestones
+		const achievedTimeMilestones = checkCodeTimeMetricsMilestonesAchieved();
+
+		// checks to see if there are any new language milestones achieved
+		const achievedLangMilestones = checkLanguageMilestonesAchieved();
+
+		// checks to see if there are any day milestones achived
+		const achievedDaysMilestones = checkDaysMilestones();
+
+		if (achievedDaysMilestones.length || achievedLangMilestones.length || achievedTimeMilestones.length) {
+			// get the current milestones so if its an update the milestone array
+			// has them all instead of getting replaced by net new milestones
+			const currentMilestones: Array<number> = getMilestonesByDate(new Date().valueOf()) || [];
+
+			// update the server
+			await this.upsertMilestones(currentMilestones);
+
+			// fetch the milestones
+			await this.fetchMilestones();
+
+			// syncs the Summary info (hours, lines, etc) to the file
+			syncSummary();
+		}
+		this._checkingForMilestones = false;
+	}
+
+	private async upsertMilestones(milestones: Array<number>) {
+		const dateNow = new Date();
+		const millisNow = dateNow.valueOf();
+		// handles creating and updating of milestones and adds milestones accordingly
+		const d = new Date(millisNow);
+		const offset_minutes = d.getTimezoneOffset();
+		const day_number = getDayNumberFromDate(millisNow);
+		const milestoneData = [{
+			day_number,
+			unix_date: Math.round(millisNow / 1000), // milliseconds --> seconds
+			local_date: Math.round(millisNow / 1000) - offset_minutes * 60, // milliseconds --> seconds,
+			offset_minutes,
+			timezone: Intl.DateTimeFormat().resolvedOptions().timeZone,
+			milestones
+		}];
+
+		await softwarePost("/100doc/milestones", milestoneData, getItem("jwt"));
+	}
+
+	/**
+ * This will return an array of..
+ * [{challenge_round, createdAt, day_number, local_date, milestones [numbers], offset_minutes, timezone, type, unix_date, userId}]
+ * @param date
+ */
+	public async fetchMilestones(date: any = null): Promise<any> {
+		const jwt = getItem("jwt");
+		const ONE_DAY_SEC = 68400000;
+		// default to today and yesterday
```

Comment:
shouldn't this get all of the milestones completed for the current challenge? otherwise, the milestone view would look like it was missing some.

Reply:
Nevermind! Seeing the function now below - as long as `fetchAllMilestones` gets called on initialize

Reply:
Yeah, i was thinking about refactoring both of those but decided to wait for now. As you said, we need to make sure we obtain the correct view of milestones but we're now using a simplified approach to fetch all then during the day fetch the day's worth. At some point we should only fetch all if there are no milestones, otherwise, as we create a milestone we should also have that on hand and not fetch it again. We can drop down the milestone fetch down to only 1 time ever :)